slackbotExercise.py

```diff
@@ -3,90 +3,280 @@
+
+HASH = "%23"
+
+
+# Configuration values to be set in setConfiguration
+class Bot:
+    def __init__(self):
+        self.setConfiguration()
+
+        self.csv_filename = "log" + time.strftime("%Y%m%d-%H%M") + ".csv"
+        self.first_run = True
+
+        # local cache of usernames
+        # maps userIds to usernames
+        self.user_cache = self.loadUserCache()
+
+        # round robin store
+        self.user_queue = []
+
+
+    def loadUserCache(self):
+        if os.path.isfile('user_cache.save'):
+            with open('user_cache.save','rb') as f:
+                self.user_cache = pickle.load(f)
+                print "Loading " + str(len(self.user_cache)) + " users from cache."
+                return self.user_cache
+
+        return {}
+
+    '''
+    Sets the configuration file.
+
+    Runs after every callout so that settings can be changed realtime
+    '''
+    def setConfiguration(self):
+        # Read variables fromt the configuration file
+        with open('config.json') as f:
+            settings = json.load(f)
+
+            self.team_domain = settings["teamDomain"]
+            self.channel_name = settings["channelName"]
+            self.min_countdown = settings["callouts"]["timeBetween"]["minTime"]
+            self.max_countdown = settings["callouts"]["timeBetween"]["maxTime"]
+            self.num_people_per_callout = settings["callouts"]["numPeople"]
+            self.sliding_window_size = settings["callouts"]["slidingWindowSize"]
+            self.group_callout_chance = settings["callouts"]["groupCalloutChance"]
+            self.channel_id = settings["channelId"]
+            self.exercises = settings["exercises"]
+
+            self.debug = settings["debug"]
+
+        self.post_URL = "https://" + self.team_domain + ".slack.com/services/hooks/slackbot?token=" + URL_TOKEN_STRING + "&channel=" + HASH + self.channel_name
+
+
+################################################################################
+'''
+Selects an active user from a list of users
+'''
+def selectUser(bot, exercise):
+    active_users = fetchActiveUsers(bot)
+
+    # Add all active users not already in the user queue
+    # Shuffles to randomly add new active users
+    shuffle(active_users)
+    bothArrays = set(active_users).intersection(bot.user_queue)
+    for user in active_users:
+        if user not in bothArrays:
+            bot.user_queue.append(user)
 
-# Set your config variables from the config.json file
-with open('config.json') as f:
-    settings = json.load(f)
-    USERTOKENSTRING = settings['USERTOKENSTRING']
-    URLTOKENSTRING = settings["URLTOKENSTRING"]
-    TEAMNAMESTRING = settings["TEAMNAMESTRING"]
+    # The max number of users we are willing to look forward
+    # to try and find a good match
+    sliding_window = bot.sliding_window_size
 
+    # find a user to draw, priority going to first in
+    for i in range(len(bot.user_queue)):
+        user = bot.user_queue[i]
```

Comment:
```suggestion
    for user in bot.user_queue:
```
I think it's better this way (: